Libs/MRML/Core/vtkMRMLScene.cxx

```diff
@@ -1350,6 +1350,7 @@ void vtkMRMLScene::RemoveNode(vtkMRMLNode *n)
     this->RemoveReferencesToNode(n);
     }
 
+  n->SetID(NULL);
```

Comment:
You are right, it should be the behavior, we just have to avoid potential regressions.
I've checked all RemoveNode calls in the Slicer source code and only found vtkMRMLScene::Undo() and vtkMRMLScene::Redo() relying on the node IDs to be preserved after RemoveNode. Fortunately, vtkMRMLScene is a friend of vtkMRMLNode, so we can save the node ID before RemoveNode and restore node ID after the call.